common/web/input-processor/src/corrections.ts

```diff
@@ -0,0 +1,93 @@
+import { CorrectionLayout } from "./correctionLayout.js";
+
+/**
+ * Computes a squared 'pseudo-distance' for the touch from each key.  (Not a proper metric.)
+ * Intended for use in generating a probability distribution over the keys based on the touch input.
+ * @param touchCoords A proportional (x, y) coordinate of the touch within the keyboard's geometry.
+ *                           Should be within <0, 0> to <1, 1>.
+ * @param correctiveLayout  The corrective-layout mappings for keys under consideration
+ *                          by a correction algorithm, also within <0, 0> to <1, 1>.
+ * @returns A mapping of key IDs to the 'squared pseudo-distance' of the touchpoint to each key.
+ */
+export function keyTouchDistances(touchCoords: {x: number, y: number}, correctiveLayout: CorrectionLayout): Map<string, number> {
+  let keyDists: Map<string, number> = new Map<string, number>();
+
+  // This double-nested loop computes a pseudo-distance for the touch from each key.  Quite useful for
```

Comment:
I don't see the double-nesting?

What is the performance of this algorithm?

Reply:
Ah, right... the refactor split that up.  The double-nesting is from `correctionLayout.ts`, `buildCorrectiveLayout` - you can see a nested `.map()` call in that method.

```suggestion
  // This loop computes a pseudo-distance for the touch from each key. Quite useful for
```

Reply:
This specific function should be O(n) - as well as the `buildCorrectiveLayout` - where `n` is the total number of keys specified in the "corrective layout".